3DSLoader

tentone opened the pull request:
Hi there

Some time ago i tried to write a 3ds loader, ended up using 3dslib as a base for this loader, named it TDSLoader.

Currently it loads only geometry without normals and uv's and loads materials properly.

Since there are some people interested in it i will leave it here, i want to come back to it some time soon to add smooth groups and uv's support, but right now i dont have the time to do it.

I hope this is usefull 

Cumps

tentone committed:
Added 3DSLoader first version

fraguada commented:
also #11010 (haven't looked at either code, just pointing out there is another commit in this direction).

tentone commented:
I was waiting a bit to finish this one and i wanted to open a pull request only when i was done!

But since there is somebody else working on it as well, i opened it up now so that it can be used to help people that are working on another 3DS loaders for threejs, or maybe people can colaborate and help finish this one sooner.

The converter made by @TheCodeCrafter its almost complete (supports pretty much everything), maybe this can be usefull to help him convert the coverter to a loader.

Cumps

TheCodeCrafter commented:
Thanks @tentone I have been looking to finish that for a while, but I haven't had the time! I already have all of the file stuff down in there, so all you have to do is match it up with the other converters, so that it is useable. That is, if you want to take up the reins of this project.

mrdoob commented:
> The converter made by @TheCodeCrafter its almost complete (supports pretty much everything), maybe this can be usefull to help him convert the coverter to a loader.

Where is that converter?

tentone commented:
@mrdoob [here](https://github.com/mrdoob/three.js/pull/11010/commits/5f37a7ffd9084aa439affd9b3807c5bee9c45038)

tentone commented:
@TheCodeCrafter I might have some time next week to take a look as this again, my target is to add UV, normals and finish some materials properties.  I think that will be enough for pretty much every case. 

Thanks a lot for the work you have done in the Converter, i will definitively take a look at it 👍

tentone committed:
Updated TDSLoader

tentone commented:
Hi

Updated the PR to include the latest version of the 3DSLoader, which already supports UV loading.

I think its ready to be used to load static geometries, has been working fine for pretty everything that i tried so far.

Cumps

mrdoob commented on examples/js/loaders/TDSLoader.js:
```diff
@@ -0,0 +1,830 @@
+"use strict";
+
+THREE.TDSLoader = function(manager)
+{
```
Do you mind changing all the curly braces so they sit on the same line?

WestLangley replied:
@mrdoob https://gist.github.com/WestLangley/604efc8fed3d9c5edfa83d4d184c6d90 Hopefully no mistakes. mrdoobapproves had a difficult time with it. :)

tentone committed:
Refactored loader and added example

tentone commented:
Thanks @WestLangley 

Updated the file (changed string with "" to '' and changed prototype declaration).
Added an example for the 3DS loader

Cumps

mrdoob commented:
Many thanks!

mrdoob commented:
Seems like the loader is not handling some transformations correctly.

![screen shot 2017-07-19 at 10 46 24](https://user-images.githubusercontent.com/97088/28381403-9b4f4574-6c6f-11e7-9b39-e82018775763.png)

![screen shot 2017-07-19 at 10 46 37](https://user-images.githubusercontent.com/97088/28381411-a1606772-6c6f-11e7-9a19-7f343b0916a7.png)

[F50N_H.3DS.zip](https://github.com/mrdoob/three.js/files/1159961/F50N_H.3DS.zip)

tentone commented:
Hi

The loader is missing a transformation parameter!

I will try to work on it this weekend.

Thanks

tentone commented:
![capture](https://user-images.githubusercontent.com/9158354/28393809-e073c2bc-6ce0-11e7-9095-ad7561baf44b.PNG)

@mrdoob 

Hi im working on this im almost done but im having a problem applying the tranformation matrix to the object, any idea of whats going on here? I will continue this tomorrow :)

Thanks :)

mrdoob commented:
`matrix[ 12 ]` is `position.x`, `matrix[ 13 ]` is `position.y` and `matrix[ 14 ]` is `position.z`;

I suspect the way you're assigning values to the `matrix.elements` is incorrect (maybe transpose?).

tentone commented:
Its not that i have to reasearch a bit more, position its already okay but rotation is not i have to check units

snowww1014 commented:
Hi, thank you for your writing TDSLoader.
Sometimes 3DS file not loaded.
I uploaded the file failed in loading.
Please help me.

[model.zip](https://github.com/mrdoob/three.js/files/1187086/model.zip)

tentone commented:
@snowww1014 
![capture](https://user-images.githubusercontent.com/9158354/28774619-4847c5ce-75e6-11e7-9e0d-9497afc641e4.JPG)

Hi i tried to load your model and it loads just fine, i will update TDSLoader today but the current version should load your file without any problem.

snowww1014 commented:
Thanks @tentone.
But I still cannot load this model.
I used the source code of examples/webgl_loader_3ds.html.

        var loader = new THREE.TextureLoader();
        var texture = loader.load( 'models/3ds/portalgun/textures/color.jpg' );
	var normal = loader.load( 'models/3ds/portalgun/textures/normal.jpg' );

	var loader = new THREE.TDSLoader( );
	loader.load( 'models/3ds/ANMOCHUANG.3ds', function ( object ) {
		object.traverse( function ( child ) {
			if ( child instanceof THREE.Mesh ) {
				child.material.map = texture;
				child.material.normalMap = normal;
			}
		} );
               scene.add( object );
	});
Best regards.

tentone commented:
@snowww1014 
I need to check your code, do you mind sending to me a .zip with a sample or posting a https://jsfiddle.net here?

Thanks a lot

snowww1014 commented:
Thanks for your help.
Uploaded source code.
Line 69 of webgl_loader_3ds.html loads the 3ds model.
Best regards.
[sample.zip](https://github.com/mrdoob/three.js/files/1190158/sample.zip)
